### PR TITLE
Initial Rails Commands Infrastructure

### PR DESCRIPTION
**Pull request by kaspth**
Adds the initial infrastructure needed to run our internal commands with Thor. It unifies the regular app commands and engine commands into one.

I've left the old command files in place in rails/commands and rails/engine/commands to slim the diff and ease review. I'll remove them before the final merge.

In another PR we can add command dependencies and port our Rake commands over.

Later, we can figure out how to let users add their own commands. It's mostly there; there's lookup code, the app commands just needs to be on the lookup path.

### CODE REVIEW

**Comment by prathamesh-sonpatki** [comment]
@kaspth Any specific reason why we have each command in its own folder like `/commands/rake/rake_command.rb`?

File: railties/lib/rails/command.rb [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+require "active_support/core_ext/enumerable"
+require "active_support/core_ext/object/blank"
+require "active_support/core_ext/hash/transform_values"
+
+require "thor"
+
+module Rails
+  module Command
+    autoload :Behavior, "rails/command/behavior"
+    autoload :Base, "rails/command/base"
```
Comment (eileencodes) [inline railties/lib/rails/command.rb]: Autoload isn't finding these without the path?

File: railties/lib/rails/command.rb [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    include Behavior
+
+    class << self
+      def hidden_commands # :nodoc:
+        @hidden_commands ||= []
+      end
+
+      def environment
```
Comment (eileencodes) [inline railties/lib/rails/command.rb]: Should this be no doc'd or documented? Looks like probably a nodoc.

File: railties/lib/rails/command.rb [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      end
+
+      def sorted_groups # :nodoc:
+        lookup!
+
+        groups = (subclasses - hidden_commands).group_by { |c| c.namespace.split(":").first }
+        groups.transform_values! { |commands| commands.flat_map(&:printing_commands).sort }
+
+        rails = groups.delete("rails")
+        [[ "rails", rails ]] + groups.sort.to_a
```
Comment (eileencodes) [inline railties/lib/rails/command.rb]: Why delete `"rails"` and then add it back?

File: railties/lib/rails/command/base.rb [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+      class << self
+        # Returns true when the app is a Rails engine.
+        def engine?
+          defined?(ENGINE_ROOT)
+        end
+
+        # Tries to get the description from a USAGE file one folder above the command
+        # root.
+        def desc(usage = nil, description = nil)
```
Comment (eileencodes) [inline railties/lib/rails/command/base.rb]: Did you call this `desc` instead of `description` because someone else defines description already? Abbreviated this way I read it was `descending`.

File: railties/lib/rails/command/environment_argument.rb [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+          if environment
+            self.options = options.merge(environment: acceptable_environment(environment))
+          end
+        end
+
+        def acceptable_environment(env = nil)
+          if available_environments.include? env
+            env
+          else
+            %w(production development test).detect { |e| e =~ /^#{env}/ } || env
```
Comment (eileencodes) [inline railties/lib/rails/command/environment_argument.rb]: I think we prefer spaces around arrays like this: `%w( production development test )...`

File: railties/lib/rails/command/base.rb (hunk 4 above)
Comment (kaspth) [inline railties/lib/rails/command/base.rb]: The `desc` is from Thor: https://github.com/erikhuda/thor/blob/06dade7c9567edccc4da2f0d6fcf4c2d00066e3d/lib/thor.rb#L54

We could add an alias for this description, which I agree would be more... descriptive 🤓

File: railties/lib/rails/command.rb (hunk 3 above)
Comment (kaspth) [inline railties/lib/rails/command.rb]: To make Rails' built in commands show up first in help output, e.g.:

```
Rails:
  console
  dbconsole
  destroy
  generate
  new
  runner
  server
  test
  version

Rake:
  about
  app:template
  app:update
  assets:clean[keep]
  assets:clobber
```

File: railties/lib/rails/command.rb (hunk 2 above)
Comment (kaspth) [inline railties/lib/rails/command.rb]: Yeah, executing a command in a given environment is probably a Rails internal thing. On the other hand having every command except a `--environment` could be neat (which this method could help assign the default for).

File: railties/lib/rails/command.rb (hunk 1 above)
Comment (kaspth) [inline railties/lib/rails/command.rb]: Oh, it works relatively? Thanks, I'll try it!

**Comment by kaspth** [comment]
@prathamesh-sonpatki mostly to make it easier to add a USAGE file or maybe other tertiary files later on.

File: railties/lib/rails/command.rb (hunk 1 above)
Comment (kaspth) [inline railties/lib/rails/command.rb]: Ruby's standard autoload wants a file mapping as well. ¯\_(ツ)_/¯

File: railties/lib/rails/command.rb (hunk 2 above)
Comment (kaspth) [inline railties/lib/rails/command.rb]: Nodoc'ing for now 👍

**Comment by kaspth** [comment]
Note that this also removes the command whitelist restrictions on the engine commands. I'll need to check the added commands more to see if they work out of the box.

E.g. it would be neat if the server command could find out if it was in an engine and then just boot `test/dummy`.

File: railties/lib/rails/command.rb (hunk 1 above)
Comment (eileencodes) [inline railties/lib/rails/command.rb]: Yea not sure how exactly that works, some places in Rails we're using it without paths, and then in other places we are using paths.  As you said ¯_(ツ)_/¯

File: railties/lib/rails/commands/server/server_command.rb [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        DoNotReverseLookup: true,
+        environment:        (ENV["RAILS_ENV"] || ENV["RACK_ENV"] || "development").dup,
+        daemonize:          false,
+        caching:            nil,
+        pid:                Options::DEFAULT_PID_PATH,
+        restart_cmd:        restart_command)
+    end
+
+    private
+
```
Comment (eileencodes) [inline railties/lib/rails/commands/server/server_command.rb]: ✂️

**Comment by maclover7** [comment]
👍 this is looking really good! Quick question just to confirm, commands override the `#perform` method to run the command, right?

File: railties/lib/rails/command.rb (hunk 1 above)
Comment (fxn) [inline railties/lib/rails/command.rb]: Because those ones use ActiveSupport::Autoload (on my phone, from memory). This is the standard in the code base, conventional file layout and no paths.

**Comment by aliAsadi92** [comment]
👍

File: railties/lib/rails/command.rb (hunk 1 above)
Comment (kaspth) [inline railties/lib/rails/command.rb]: Oh right, I'll pull that module in and save those paths 😁

**Comment by kaspth** [comment]
@maclover7 not sure what you mean? The instance level `perform` is just added as an alias for the actual command name, e.g. the test command would want an instance level `test` method. But I found that to be too much repetition for our internal common case of one command per file.

So yes, right now they're one command per file. But it shouldn't have to be that way, we just don't have to tackle it upfront and let the infrastructure as we solve more use cases. This is a starting point 😁

File: railties/lib/rails/command/behavior.rb [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        # Track all command subclasses.
+        def subclasses
+          @subclasses ||= []
+        end
+
+        protected
+
+          # This code is based directly on the Text gem implementation.
+          # Copyright (c) 2006-2013 Paul Battley, Michael Neumann, Tim Fletcher.
+          #
```
Comment (kaspth) [inline railties/lib/rails/command/behavior.rb]: Pssst @matthewd

**Comment by kaspth** [comment]
Removed the old command files 😁

File: railties/lib/rails/command/behavior.rb [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+          # Returns a value representing the "cost" of transforming str1 into str2
```
Comment (rafaelfranca) [inline railties/lib/rails/command/behavior.rb]: `.` in the end.

File: railties/lib/rails/commands/server/server_command.rb [hunk 9]
```diff
@@ -88,9 +90,7 @@ def middleware
     end
 
     def default_options
-      super.merge(
-        Port:               ENV.fetch("PORT", 3000).to_i,
-        Host:               ENV.fetch("HOST", "localhost").dup,
+      super.merge(        Port:               ENV.fetch("PORT", 3000).to_i,
```
Comment (rafaelfranca) [inline railties/lib/rails/commands/server/server_command.rb]: Why Host was removed? and also the style is wrong here.

File: railties/lib/rails/test_unit/minitest_plugin.rb [hunk 10]
```diff
@@ -1,4 +1,5 @@
 require "active_support/core_ext/module/attribute_accessors"
+require "active_support/core_ext/hash/keys"
```
Comment (rafaelfranca) [inline railties/lib/rails/test_unit/minitest_plugin.rb]: Why is this here?

**Review by rafaelfranca — CHANGES_REQUESTED** [review]

**Comment by rafaelfranca** [comment]
Tests are also broken.

File: railties/lib/rails/test_unit/minitest_plugin.rb (hunk 10 above)
Comment (kaspth) [inline railties/lib/rails/test_unit/minitest_plugin.rb]: Some leftover from an earlier change.

File: railties/lib/rails/commands/server/server_command.rb (hunk 9 above)
Comment (kaspth) [inline railties/lib/rails/commands/server/server_command.rb]: A fault from a rebase I think. Fixed, thanks!

**Comment by kaspth** [comment]
Got the build back to green again. @rafaelfranca since you requested the last changes, what do you say?

File: railties/lib/rails/command/behavior.rb [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+          @subclasses ||= []
+        end
+
+        protected
+
+          # This code is based directly on the Text gem implementation.
+          # Copyright (c) 2006-2013 Paul Battley, Michael Neumann, Tim Fletcher.
+          #
+          # Returns a value representing the "cost" of transforming str1 into str2.
+          def levenshtein_distance(str1, str2)
```
Comment (andremedeiros) [inline railties/lib/rails/command/behavior.rb]: I could totally see a use for this in the ActiveSupport string extensions.

Comment (kaspth) [inline railties/lib/rails/command/behavior.rb]: I'm not sure we want to make this a public API. Maybe as `edit_distance_to('other string')`, but when would you want the distance directly versus just sort an array by it?

E.g. this might be more suitable:

```ruby
%w( some words with a distance ).sort_by_edit_distances_to('this string')
```

Comment (andremedeiros) [inline railties/lib/rails/command/behavior.rb]: Sounds like the former would totally back up the latter here, right? Or am I missing something?

Comment (kaspth) [inline railties/lib/rails/command/behavior.rb]: Yes, but I'm trying if and how we'd want to expose this as an API :)

Comment (kaspth) [inline railties/lib/rails/command/behavior.rb]: Spoke with the team and we're not gonna provide this as a public API. But maybe you can try the did_you_mean gem? It's bundled with Ruby 2.3 and also includes this code: https://github.com/yuki24/did_you_mean/blob/c22ef694f411ed4f656a029630847129c78c4cb1/lib/did_you_mean/levenshtein.rb

File: railties/lib/rails/command.rb (hunk 3 above)
Comment (RichardBlair) [inline railties/lib/rails/command.rb]: In that case it might make more sense to remove rails from the groups, group the commands, transform, and then prepend rails?

Comment (RichardBlair) [inline railties/lib/rails/command.rb]: Sorry, just realized this was merged. I can just follow up with a PR.
